Add support for a 'at' collection listRecords filter.

n1k0 opened the pull request:
This adds a new `at` option to `Collection#listRecords` to retrieve a snapshot of the collection records at a given timestamp in the past.

- [x] Compute collection list of records state at a given timestamp back in time
- [x] Ensure paginating long lists of changes
- [x] Update documentation
- [ ] Fix travis build setup

f=? @leplatrem

n1k0 commented:
TGIF

![](http://i.imgur.com/nHi6FC1.png)

![computing](https://cloud.githubusercontent.com/assets/41547/23315261/cf5f9f56-fac5-11e6-9fc1-431b08526912.gif)

n1k0 commented:
There's a bug in Kinto history that prevent paginating list of changes it seems: https://github.com/Kinto/kinto/issues/1114

leplatrem commented on src/collection.js:
```diff
@@ -322,7 +323,38 @@ export default class Collection {
+            snapshot.push(record);
+          }
+        }
+        return snapshot.sort((a, b) => b.last_modified - a.last_modified);
```
nit: should we respect the `sort` option provided in `listRecords()`?

n1k0 replied:
That's a good idea, I'm just slightly afraid that would open the door to supporting all other options. I liked the opinionated simplistic approach here, but we may want to think about it further.

leplatrem commented on test/integration_test.js:
```diff
@@ -1430,6 +1430,56 @@ describe("Integration tests", function() {
               });
             });
 
+            describe("'at' retrieves a snapshot at a given timestam", () => {
```
typo: *timestamp*

leplatrem commented on test/integration_test.js:
```diff
@@ -1430,6 +1430,56 @@ describe("Integration tests", function() {
+              let rec1, rec2, rec3;
+
+              beforeEach(() => {
+                return coll.listRecords()
```
nit: why do we list first?

n1k0 replied:
A leftover most probably, will remove.

leplatrem commented on test/integration_test.js:
```diff
@@ -1430,6 +1430,56 @@ describe("Integration tests", function() {
+
+              it("should handle creations", () => {
+                return coll.listRecords({at: rec1.last_modified})
+                  .should.eventually.become([
```
I think it should return the same structure as `_paginatedList` : 
https://github.com/Kinto/kinto-http.js/blob/master/src/base.js#L410-L415

n1k0 replied:
So you'd want to allow paginating a snapshot?

n1k0 replied:
Poke, I'd really like to know about that, and also if you have a plan in mind to achieve it simply :)

n1k0 replied:
Or would something like this be good enough?

```js
        return {
          last_modified: Math.max.apply(null, snapshot.map(r => r.las_modified)),
          data: snapshot.sort((a, b) => b.last_modified - a.last_modified),
          next: () => { throw new Error("Snapshots don't support pagination"); },
          hasNextPage: false,
          totalRecords: snapshot.length,
        };
```

leplatrem commented on src/collection.js:
```diff
@@ -322,7 +323,38 @@ export default class Collection {
+    if (!Number.isInteger(at) || at <= 0) {
+      throw new Error("Invalid argument, expected a positive integer.");
+    }
+    return this.bucket.listHistory({
```
We could add a XXX comment  that explains that we take the history forward, and that it would more efficient (and more chance to be complete) if we'd do it backward

n1k0 replied:
We could also update the code to process the history backward directly if that would prevents later issues.

n1k0 commented:
> I find it a lot easier to handle than on the server... No?

Having a server side implementation is much more of a strategic move then a technical one imho, but as you highligted in previous comments if the kinto-admin is the only project that would benefit from it it indeed challenges the "strategy" quite a lot ;)

Natim commented:
To be honest my call on making it a server feature was to move forward the issue about syncing between Kinto and CouchDB. Issue that was brought to us by CozyCloud.

n1k0 commented:
I think I've addressed most comments in https://github.com/Kinto/kinto-http.js/pull/161/commits/31674f78d4d30ed591d897ae6d62b1c7ee606609. We could always iterate further in next PRs to improve the implementation.

New feedback requested @leplatrem @Natim

leplatrem commented on src/collection.js:
```diff
@@ -322,7 +323,51 @@ export default class Collection {
+          }
+        }
+        return {
+          last_modified: Math.max.apply(null, snapshot.map(r => r.last_modified)),
```
A priori, this would just be `last_modified: String(at)` ?

n1k0 replied:
Not necessarily as the `at` parameter may be higher than the latest record timestamp. I felt more accurate to rely on the actual data. Also, the collection last_modified is supposed to be a number I think.

leplatrem replied:
@n1k0 I realize this would give a numeric value, whereas we want an ETag. Just use `String(at)` and you're good

n1k0 replied:
Wow I realize we expose the ETag in the list result object in kinto-http, while Kinto exposes `last_modified` as an integer in the collection attributes. This is rather confusing tbh. Oh well, will update as suggested.

n1k0 replied:
Done in https://github.com/Kinto/kinto-http.js/pull/161/commits/ce8a1c763664906ce85afb91f37e912a0488a061

n1k0 commented:
Cool, now I "just" need to fix the travis build setup, while I can't access the build logs anymore (aws issue still ongoing?)

n1k0 commented:
I'm landing this.
